### ISSUE TITLE
Mask mode sometimes resumes incorrectly

### ISSUE THREAD

**Issue by magnumripper** [magnumripper/JohnTheRipper#2306]
See recent comments to #2022 for some discussion (although it's probably unrelated).

I have yet to find a reasonable way to trigger the problem, it normally works fine. But recently I saw it, in a scenario with many salts and a very high GWS (keys-per-crypt) in OpenCL, resulting in tens of minutes for completing a batch for all salts (the bcrypt leak of last year should have the same problem).
```
$ ../run/john -form:md5crypt-opencl md5crypt_test.in -mask:?d -min-len=5 -max-len=8
Device 0: GeForce GTX TITAN X
Using default input encoding: UTF-8
Loaded 13714 password hashes with 13714 different salts (md5crypt-opencl, crypt(3) $1$ [MD5 OpenCL])
Remaining 13697 password hashes with 13697 different salts
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:29:46  (7) 0g/s 0p/s 4134Kc/s 4134KC/s GPU:81°C 11111..3562491
Session aborted
```
So far so good. Note the 0p/s which tells us not even the very first batch of keys was completed for every salt.

When resuming:
```
$ ../run/john -res
Device 0: GeForce GTX TITAN X
Loaded 13714 password hashes with 13714 different salts (md5crypt-opencl, crypt(3) $1$ [MD5 OpenCL])
Remaining 13696 password hashes with 13696 different salts
Proceeding with mask:?d
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:29:47  (8) 0g/s 0p/s 4132Kc/s 4132KC/s GPU:61°C 01111111
0g 0:00:29:58  (8) 0g/s 0p/s 4140Kc/s 4140KC/s GPU:69°C 01111111..66220111
Session aborted
```
That ended up completely fcuked up - it took off at next length. I should try to reproduce this and back up the session files before resuming.

Note that this format is not GPU-side mask, and also I didn't use node/fork/MPI. Just a soft abort (letting it finish cleanly) and then a resume.

Re-starting the job from scratch and I didn't see the same happen(!). A difference though, was that the auto-tune happened to end up with a way lower GWS (or keys-per-crypt) for whatever reason.

**Comment by magnumripper** [comment]
This is what had me pay attention to #2114. Perhaps we should try to rebase that one (still in the/a topic branch). See if we can reliably reproduce this problem, and then see if #2114 fixes it.

**Comment by magnumripper** [comment]
Hmmm come to think of it, this isn't even hybrid mask. I believe #2114 is mostly about hybrid mask.

So this should be pretty much as simple as John Proper's resuming. I can't see how/why it would ever end up wrong, let alone intermittently.

**Comment by claudioandre** [comment]
> A difference though, was that the auto-tune happened to end up with a way lower GWS (or keys-per-crypt) for whatever reason.

Have you tried to force a high GWS to see if you can reproduce the error?

**Comment by magnumripper** [comment]
Hopefully nailed and fixed in 030795f, after finding and fixing another bug in 9b66341
